File: darts-scorer-backend/src/main/java/com/dartscorer/service/GameService.java
package com.dartscorer.service;

import com.dartscorer.model.*;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.NotFoundException;
import jakarta.ws.rs.BadRequestException;
import java.util.List;

/**
 * Service for managing game logic
 */
@ApplicationScoped
public class GameService {
    
    /**
     * Create a new game with players
     */
    @Transactional
    public Game createGame(GameMode gameMode, List<String> playerNames) {
        if (playerNames == null || playerNames.isEmpty()) {
            throw new BadRequestException("At least one player is required");
        }
        
        if (playerNames.size() > 8) {
            throw new BadRequestException("Maximum 8 players allowed");
        }
        
        // Create game
        Game game = new Game(gameMode);
        
        // Create or find players and add to game
        for (String playerName : playerNames) {
            Player player = Player.findByName(playerName);
            if (player == null) {
                player = new Player(playerName);
                player.persist();
            }
            game.addPlayer(player);
        }
        
        // Start the game
        game.start();
        game.persist();
        
        // Create first turn for first player
        Turn firstTurn = game.createNewTurn();
        firstTurn.persist();
        
        return game;
    }
    
    /**
     * Get game by ID
     */
    public Game getGame(Long gameId) {
        Game game = Game.findById(gameId);
        if (game == null) {
            throw new NotFoundException("Game not found");
        }
        return game;
    }
    
    /**
     * Record a throw in the current turn
     */
    @Transactional
    public Game recordThrow(Long gameId, Integer sector, Integer multiplier) {
        Game game = getGame(gameId);
        
        if (!game.isInProgress()) {
            throw new BadRequestException("Game is not in progress");
        }
        
        // Validate input
        // Sector 0 is allowed for miss (no score)
        if (sector < 0 || sector > 25) {
            throw new BadRequestException("Invalid sector: must be between 0 and 25");
        }
        
        // For miss (sector 0), multiplier must be 0
        if (sector == 0 && multiplier != 0) {
            throw new BadRequestException("Miss (sector 0) must have multiplier 0");
        }
        
        // For non-miss throws, validate multiplier
        if (sector > 0 && (multiplier < 1 || multiplier > 3)) {
            throw new BadRequestException("Invalid multiplier: must be 1, 2, or 3");
        }
        
        // Bull (25) can only be single (1) or double (2)
        if (sector == 25 && multiplier == 3) {
            throw new BadRequestException("Bull cannot be triple");
        }
        
        // Get or create current turn
        Turn currentTurn = game.getCurrentTurn();
        if (currentTurn == null) {
            currentTurn = game.createNewTurn();
            currentTurn.persist();
        }
        
        // Check if turn is already complete
        if (currentTurn.isComplete()) {
            throw new BadRequestException("Current turn is already complete. Call nextPlayer first.");
        }
        
        // Create and add throw
        int throwNumber = currentTurn.getThrowCount() + 1;
        Throw dartThrow = new Throw(currentTurn, sector, multiplier, throwNumber);
        dartThrow.persist();
        
        currentTurn.addThrow(dartThrow);
        
        // Process throw based on game mode
        processThrow(game, currentTurn, dartThrow);
        
        return game;
    }
    
    /**
     * Process a throw based on game mode
     */
    private void processThrow(Game game, Turn turn, Throw dartThrow) {
        switch (game.gameMode) {
            case TRAINING:
                // In training mode, just accumulate score
                break;
                
            case STANDARD_301:
            case DOUBLE_OUT_301:
                process301Throw(game, turn, dartThrow);
                break;
        }
    }
    
    /**
     * Process a throw in 301 mode
     */
    private void process301Throw(Game game, Turn turn, Throw dartThrow) {
        // Get current score for player
        Integer currentScore = game.getPlayerScore(turn.player);
        
        // Calculate new score after this throw
        Integer newScore = currentScore - dartThrow.score;
        
        // Check for bust (went below 0 or exactly 0 without proper finish)
        if (newScore < 0) {
            turn.isBust = true;
            turn.remainingScore = currentScore; // Score remains unchanged
            return;
        }
        
        // Check for exact finish
        if (newScore == 0) {
            // In DOUBLE_OUT mode, must finish with a double
            if (game.gameMode == GameMode.DOUBLE_OUT_301 && !dartThrow.isDouble()) {
                turn.isBust = true;
                turn.remainingScore = currentScore;
                return;
            }
            
            // Valid finish!
            turn.remainingScore = 0;
            game.complete(turn.player);
            return;
        }
        
        // Valid throw, update remaining score
        turn.remainingScore = newScore;
    }
    
    /**
     * Move to next player
     */
    @Transactional
    public Game nextPlayer(Long gameId) {
        Game game = getGame(gameId);
        
        if (!game.isInProgress()) {
            throw new BadRequestException("Game is not in progress");
        }
        
        // Get the last turn for the current player
        Turn lastTurn = game.turns.stream()
            .filter(t -> t.player.id.equals(game.getCurrentPlayer().id))
            .reduce((first, second) -> second)
            .orElse(null);
        
        // Check if the last turn is complete (has 3 throws)
        if (lastTurn == null || !lastTurn.isComplete()) {
            throw new BadRequestException("Current turn is not complete");
        }
        
        // Move to next player
        game.nextPlayer();
        
        // Create new turn for next player
        Turn nextTurn = game.createNewTurn();
        
        // Set initial remaining score for 301 modes
        if (game.gameMode != GameMode.TRAINING) {
            nextTurn.remainingScore = game.getPlayerScore(game.getCurrentPlayer());
        }
        
        nextTurn.persist();
        
        return game;
    }
    
    /**
     * Delete a game
     */
    @Transactional
    public void deleteGame(Long gameId) {
        Game game = getGame(gameId);
        game.delete();
    }
    
    /**
     * Get all active games
     */
    public List<Game> getActiveGames() {
        return Game.findActiveGames();
    }
    
    /**
     * Get game history (recent turns)
     */
    public List<Turn> getGameHistory(Long gameId) {
        Game game = getGame(gameId);
        return game.turns;
    }
}

File: darts-scorer-frontend/mobile-backend/services/GameService.js
const Game = require('../models/Game');
const Turn = require('../models/Turn');

class GameService {
  createGame(gameMode, playerNames) {
    return Game.create(gameMode, playerNames);
  }

  getGame(gameId) {
    const game = Game.findById(gameId);
    if (!game) {
      throw new Error('Gioco non trovato');
    }
    return game;
  }

  recordThrow(gameId, sector, multiplier) {
    const game = this.getGame(gameId);
    
    if (!game.isInProgress()) {
      throw new Error('Il gioco non è in corso');
    }

    // Validazione input
    if (sector < 0 || sector > 25) {
      throw new Error('Settore non valido: deve essere tra 0 e 25');
    }

    // Per miss (sector 0), multiplier deve essere 0
    if (sector === 0 && multiplier !== 0) {
      throw new Error('Miss (settore 0) deve avere moltiplicatore 0');
    }

    // Per lanci non-miss, valida il moltiplicatore
    if (sector > 0 && (multiplier < 1 || multiplier > 3)) {
      throw new Error('Moltiplicatore non valido: deve essere 1, 2, o 3');
    }

    // Bull (25) può essere solo singolo (1) o doppio (2)
    if (sector === 25 && multiplier === 3) {
      throw new Error('Bull non può essere triplo');
    }

    // Ottieni o crea il turno corrente
    let currentTurn = game.getCurrentTurn();
    if (!currentTurn) {
      currentTurn = game.createNewTurn();
    }

    // Verifica se il turno è già completo
    if (currentTurn.isComplete()) {
      throw new Error('Turno corrente già completo. Chiama nextPlayer prima.');
    }

    // Aggiungi il lancio
    const dartThrow = currentTurn.addThrow(sector, multiplier);

    // Processa il lancio in base alla modalità di gioco
    this.processThrow(game, currentTurn, dartThrow);

    return game;
  }

  processThrow(game, turn, dartThrow) {
    switch (game.gameMode) {
      case Game.GameMode.TRAINING:
        // In modalità training, accumula solo il punteggio
        break;
        
      case Game.GameMode.STANDARD_301:
      case Game.GameMode.DOUBLE_OUT_301:
        this.process301Throw(game, turn, dartThrow);
        break;
    }
  }

  process301Throw(game, turn, dartThrow) {
    // Ottieni il punteggio corrente del giocatore
    const currentScore = game.getPlayerScore(turn.playerId);
    
    // Calcola il nuovo punteggio dopo questo lancio
    const newScore = currentScore - dartThrow.score;
    
    // Verifica bust (andato sotto 0 o esattamente 0 senza finish corretto)
    if (newScore < 0) {
      turn.setBust(true);
      turn.updateRemainingScore(currentScore); // Il punteggio rimane invariato
      return;
    }
    
    // Verifica finish esatto
    if (newScore === 0) {
      // In modalità DOUBLE_OUT, deve finire con un doppio
      if (game.gameMode === Game.GameMode.DOUBLE_OUT_301 && !dartThrow.isDouble()) {
        turn.setBust(true);
        turn.updateRemainingScore(currentScore);
        return;
      }
      
      // Finish valido!
      turn.updateRemainingScore(0);
      game.complete(turn.playerId);
      return;
    }
    
    // Lancio valido, aggiorna il punteggio rimanente
    turn.updateRemainingScore(newScore);
  }

  nextPlayer(gameId) {
    const game = this.getGame(gameId);
    
    if (!game.isInProgress()) {
      throw new Error('Il gioco non è in corso');
    }

    // Ottieni l'ultimo turno del giocatore corrente
    const currentPlayer = game.getCurrentPlayer();
    const turns = Turn.findByGameAndPlayer(game.id, currentPlayer.id);
    const lastTurn = turns[turns.length - 1];

    // Verifica che l'ultimo turno sia completo (ha 3 lanci)
    if (!lastTurn || !lastTurn.isComplete()) {
      throw new Error('Turno corrente non completo');
    }

    // Passa al prossimo giocatore
    game.nextPlayer();

    return game;
  }

  deleteGame(gameId) {
    const game = this.getGame(gameId);
    game.delete();
  }

  getActiveGames() {
    return Game.findActiveGames();
  }

  getGameHistory(gameId) {
    const game = this.getGame(gameId);
    return game.getTurns();
  }
}

module.exports = new GameService();

// Made with Bob
